keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1188,239 +937,309 @@ drivers are expected to be written for Kubernetes.
+// must only use device definitions from the slices with the highest generation number
+// and ignore the others.
+//
+// If necessary, a consumer can check the number of total devices in a pool (included
+// in each slice) to determine whether its view of a pool is complete.
+//
+// For devices that are not local to a node, the node name is not set. Instead,
+// a pool name is chosen by the driver if necessary. It can be left blank if the
+// device names are unique in the entire cluster. The unique identifier then
+// has the same format as before with `<driver name>[/<pool name>]/<device name>`.
 type ResourceSlice struct {
     metav1.TypeMeta
     // Standard object metadata
+    // +optional
     metav1.ObjectMeta
 
-    // NodeName identifies the node which provides the resources
-    // if they are local to a node.
+    Spec ResourceSliceSpec
+
+    // A status might get added later.
+}
+
+type ResourceSliceSpec struct {
+    // NodeName identifies the node which provides the devices. All devices
+    // are local to that node and part of a per-node pool.
     //
-    // A field selector can be used to list only ResourceSlice
-    // objects with a certain node name.
+    // This field can be used to limit access from nodes to slices with
+    // the same node name. It also indicates to autoscalers that adding
+    // new nodes of the same type as some old node might also make new
+    // devices available.
+    //
+    // NodeName and PoolName are mutually exclusive.
+    //
+    // +optional
     NodeName string
 
+    // PoolName can be used to describe pools of devices which are not local
+    // to a node. It can be left blank if the device names are unique in the cluster.
+    //
+    // It must not be longer than 253 and must consist of one or more DNS sub-domains
+    // separated by slashes.
+    //
+    // +optional
+    PoolName string
+
+    // The generation gets bumped in all slices of a pool whenever device
+    // definitions changed. A consumer must only use device definitions from the slices
+    // with the highest generation number and ignore the others.
+    PoolGeneration int64
+
+    // Defines which nodes have access to the devices in the pool.
+    // Must not be set when the node name is set.
+    //
+    // If the node name is not set and this field is unset, devices
+    // are accessible in the entire cluster.
+    //
+    // +optional
+    NodeSelector *v1.NodeSelector
+
     // DriverName identifies the DRA driver providing the capacity information.
     // A field selector can be used to list only ResourceSlice
     // objects with a certain driver name.
+    //
+    // Must be a DNS subdomain and should end with a DNS domain owned by the
+    // vendor of the driver.
     DriverName string
 
-    ResourceModel
+    // The total number of devices in the pool.
+    // Consumers can use this to check whether they have
+    // seen all slices.
+    TotalDeviceNumber int64
```

Comment:
Would `TotalDeviceCount` be a better name?

Reply:
Yes.

Reply:
I don't think it technically matters, but maybe renaming this to `PoolDeviceCount` would be appropriate. We have `PoolName`, `PoolGeneration`, etc. and now this is the count of devices in that pool.

Reply:
Let's use `PoolDeviceCount`. There's a nice symmetry to it.

I prefer the count of devices over count of slices because it is information that may be useful to a user when looking at one slice. The number of slices is less interesting and also could change more often.